### DISCUSSION TITLE
How to specify tidy options?

### DISCUSSION THREAD

**Discussion by Ifitrains** [quarto-dev/quarto-cli#5813] [Q&A] [answered]
From this [question](https://stackoverflow.com/questions/76406452/horizontal-code-overflow-in-quarto-revealjs), I was wondering how to use tidy options in the new chunk options, I thought something along these lines could work, but it does not: 

```
---
format: revealjs
---

## A title


Here is some example code


```{r}
#| eval: false
#| echo: true
#| tidy: true
#| tidy-opts: 
#|   - width-cutoff=60
multipoint <- sf::st_multipoint(rbind(c(3.2,4), c(3,4.6), c(3.8,4.4), c(3.5,3.8), c(3.4,3.6), c(3.9,4.5)))
```
```

**Comment by mcanouil** [comment, marked as answer]
There is no "=" in YAML, it's ":".

To define it in code cell:

````qmd
```{r}
#| eval: false
#| echo: true
#| tidy: true
#| tidy.opts: 
#|   - width.cutoff: 60
```
````

Or

````qmd
```{r}
#| eval: false
#| echo: true
#| tidy: true
#| tidy.opts: { width.cutoff: 60 }
```
````

~~If the option with the "dash" notation does not work (it should I think), try with the "dot".~~

To define it globally, see <https://quarto.org/docs/computations/r.html#knitr-options>.

**Reply by Ifitrains** [reply]
Thank you for your quick response! Your suggestions do not work however, one get the following error for both solutions:

> ✖ The value width.cutoff: 60 is of type object.

**Reply by mcanouil** [reply]
> If the option with the "dash" notation does not work (it should I think), try with the "dot".

It works as suggested with the dot notation.
The dash version is not a Quarto option thus does not exist.

**Reply by Ifitrains** [reply]
Could you post it maybe as a full answer? I do not really get what you mean, sorry!

**Reply by mcanouil** [reply]
I just edited my answer.

**Reply by Ifitrains** [reply]
the second one works for me, many thanks! :)

**Reply by Ifitrains** [reply]
Then `tidy-opts` should also be fixed in the documentation: https://quarto.org/docs/reference/cells/cells-knitr.html#code-output

**Reply by cscheid** [reply]
> Then `tidy-opts` should also be fixed in the documentation: https://quarto.org/docs/reference/cells/cells-knitr.html#code-output

For future readers of this thread, please be aware that the options in the reference documentation are not exhaustive. It's impossible to cover every possible expectation of downstream libraries, because that code can read the metadata however it wants. In other words, this isn't a gap in the reference documentation.

It might be possible to fix this in our narrative docs, but it won't show up in the `docs/reference` section.

**Reply by mcanouil** [reply]
@cscheid I think there is something fishy here. `tidy-opts` is listed in reference docs, is listed in the options to be normalised, _i.e._, it's converted to `tidy.opts`, but somehow there is something getting in the way leading to an error.
I can take a deeper look into this.

In some aspects, it's related to #5744 and other issues regarding "dash" versus "dot" notations.
There is, in my opinion room, for enhancement to reduce cognitive hassle between the two notations.

@cderv in the eventuality you already have some insights 😏

**Reply by cderv** [reply]
I believe this is the hint of an issue in our normalization process and how options are (or rather should be) replaced. 

1. **knitr** defines `tidy.opts = NULL` by default
2. Setting `tidy-opts` will be parsed by **knitr** as `tidy-opts` and will not replace the default one registered at `tidy.opts`
3. Normalization is done by Quarto as an option hooks. This means options are process and `tidy-opts` is indeed changed to `tidy.opts`. 
4. But not options in **knitr** have twice `tidy.opts` and only the first one will apply (`tidy.opts = NULL`). 

So our normalization is buggy ! Doing `tidy.opts` directly in a chunk will not have the issue because at the end **knitr** parses the options anyway. 

So 

* we need to fix the normalization process to **replace** options in Quarto
* We probably need to make **knitr** do the normalization itself at some point **when parsing the YAML chunk options**

**Reply by mcanouil** [reply]
IMHO, with the YAML style in knitr, It would make sense to me, that knitr natively supports "dash" notations itself.

**Reply by cderv** [reply]
Issue opened to track this : https://github.com/quarto-dev/quarto-cli/issues/5852

thanks everyone ! We would not have found this easily !

**Reply by mcanouil** [reply]
Closing this in favour of the issue.

**Comment by cderv** [comment]
FWIW  from next knitr version 1.44 (and from current dev version 1.43.7), all dash options are changed to their dot versions. This includes `tidy-opts` which means it can be used in YAML and correctly replace knitr's defaults
